README
NAME
    Perinci::Sub::Property::result_postfilter - (DEPRECATED) Postfilter
    function result

VERSION
    version 0.08

SYNOPSIS
     # in function metadata
     result_postfilter => {
         re   => 'str',    # convert regex object to string
         date => 'epoch',  # convert date object to an integer Unix time
         code => 'str',    # convert subroutine to string literal 'CODE'
     }

DESCRIPTION
    NOTE: The use of this property is now deprecated. Generating filtering
    code for each function is quite wasteful when there are hundreds or more
    functions that are wrapped. Instead, see Data::Clean::JSON.

    This property specifies postfilters for function result. Currently the
    focus of this property is converting values that are unsafe when
    exporting to JSON/YAML.

    Value: a hash containing instruction of how to filter. Known keys:

    *   "re" (filter regex object)

        With filter values: "str" (stringify it).

    *   "date" (date object, such as DateTime in Perl)

        With filter values: "epoch" (convert to Unix epoch time).

    *   "code" (subroutine)

        With filter values: "str" (convert to string literal "CODE").

    More sophisticated filtering rules might be specified in the future.

    Filtering using generated code can be faster since the code will use
    for() loop and inline conversion instead of callback for each data item
    (higher subroutine call overhead).

NOTES
    Another alternative is using the encoder's facility, for example the
    JSON module can convert blessed object:

     use JSON 2;
     my $encoder = JSON->new->convert_blessed;
     {
         local *DateTime::TO_JSON = sub { $_[0]->ymd };
         print $encoder->encode($doc);
     }

    But this currently can't convert coderefs. JSON also can't handle
    circular references, which neither this wrapper property nor the above
    way can work around.

SEE ALSO
    Perinci

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/Perinci-Sub-Property-result_postfilter>.

SOURCE
    Source repository is at
    <https://github.com/sharyanto/perl-Perinci-Sub-Property-result_postfilte
    r>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=Perinci-Sub-Property-
    result_postfilter>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perinci-Sub-Property-result_postfilter

0.08     2014-03-28 (SHARYANTO)

         - No functional changes.

         - Adjust to Perinci::Sub::Wrapper 0.57 (changed section names).


0.07     2012-12-21 (SHARYANTO)

         - No functional changes.

         - Use Perinci::Sub::PropertyUtil.


0.06     2012-11-27 (SHARYANTO)

         - No functional changes.

         - Correct namespace.


0.05     2012-08-09 (SHARYANTO)

         - No functional changes.

         - Correct casing in name: Perinci::Sub::{p,P}roperty::


0.04     2012-08-08 (SHARYANTO)

         - No functional changes.

         - Mention Data::Clean::JSON.


0.03     2012-08-08 (SHARYANTO)

         - No functional changes.

         - Mark deprecated.


0.02     2012-07-31 (SHARYANTO)

         - No functional changes.

         - Update to wrapping protocol version 2.


0.01     2012-02-29 (SHARYANTO)

         - First release.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "DateTime" => 0,
    "Module::Build" => "0.3601",
    "Perinci::Sub::Wrapper" => "0.30",
    "Test::More" => "0.96",
    "Test::Perinci::Sub::Wrapper" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "(DEPRECATED) Postfilter function result",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Perinci-Sub-Property-result_postfilter",
  "dist_version" => "0.07",
  "license" => "perl",
  "module_name" => "Perinci::Sub::Property::result_postfilter",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Perinci::Sub::PropertyUtil" => 0,
    "perl" => "5.010000"
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Perinci/Sub/Property/result_postfilter.pm
package Perinci::Sub::Property::result_postfilter;

use 5.010001;
use strict;
use warnings;

use Perinci::Sub::PropertyUtil qw(declare_property);

our $VERSION = '0.08'; # VERSION

sub filter_using_for {
    my ($self, %args) = @_;

    my $v    = $args{new} // $args{value};
    return unless $v && keys(%$v);

    $self->select_section('after_call_after_res_validation');
    $self->push_lines('', '# postfilter result');

    my $term = $self->{_meta}{result_naked} ? '$_w_res' : '$_w_res->[2]';

    my $errp = "result_postfilter: Unknown filter";
    my $gen_process_item = sub {
        my $t = shift;
        my $code = '';
        while (my ($a, $b) = each %$v) {
            if ($a eq 're') {
                $code .= ($code ? "elsif":"if").
                    "(ref($t) eq 'Regexp'){";
                if ($b eq 'str') {
                    $code .= "$t = \"$t\"";
                } else {
                    die "$errp for $a: $b";
                }
                $code.="}";
            } elsif ($a eq 'date') {
                $code .= ($code ? "elsif":"if").
                    "(ref($t) eq 'DateTime'){";
                if ($b eq 'epoch') {
                    $code .= "$t = $t->epoch";
                } else {
                    die "$errp for $a: $b";
                }
                $code.="}";
            } elsif ($a eq 'code') {
                $code .= ($code ? "elsif":"if").
                    "(ref($t) eq 'CODE'){";
                if ($b eq 'str') {
                    $code .= "$t = 'CODE'";
                } else {
                    die "$errp for $a: $b";
                }
                $code.="}";
            } else {
                die "$errp $a";
            }
        }
        $code .= "elsif(ref($t) eq 'ARRAY') { \$_w_resf_ary->($t) }";
        $code .= "elsif(ref($t) eq 'HASH') { \$_w_resf_hash->($t) }";
        $code;
    };

    $self->push_lines(
        'state $_w_resf_ary; state $_w_resf_hash;');
    $self->push_lines(
        'if (!$_w_resf_ary ) { $_w_resf_ary  = sub { '.
            'for my $el (@{$_[0]}) { '.
                $gen_process_item->('$el') . ' } } }'
            );
    $self->push_lines(
        'if (!$_w_resf_hash) { $_w_resf_hash = sub { '.
            'my $h=shift; for my $k (keys %$h) { '.
                $gen_process_item->('$h->{$k}') . ' } } }'
            );
    $self->push_lines(
        "for ($term) { " . $gen_process_item->('$_') . " }",
    );
}

our $_implementation = 'for';

declare_property(
    name => 'result_postfilter',
    type => 'function',
    schema => ['hash'],
    wrapper => {
        meta => {
            v       => 2,
            # very low, we want to be the last to process result before
            # returning it
            prio    => 100,
            convert => 1,
        },
        handler => sub {
            if ($_implementation eq 'for') {
                filter_using_for(@_);
            } else {
                die "Unknown implementation '$_implementation'";
            }
        },
    },
);

1;
# ABSTRACT: (DEPRECATED) Postfilter function result

__END__

=pod

=encoding UTF-8

=head1 NAME

Perinci::Sub::Property::result_postfilter - (DEPRECATED) Postfilter function result

=head1 VERSION

version 0.08

=head1 SYNOPSIS

 # in function metadata
 result_postfilter => {
     re   => 'str',    # convert regex object to string
     date => 'epoch',  # convert date object to an integer Unix time
     code => 'str',    # convert subroutine to string literal 'CODE'
 }

=head1 DESCRIPTION

B<NOTE:> The use of this property is now deprecated. Generating filtering code
for each function is quite wasteful when there are hundreds or more functions
that are wrapped. Instead, see L<Data::Clean::JSON>.

This property specifies postfilters for function result. Currently the focus of
this property is converting values that are unsafe when exporting to JSON/YAML.

Value: a hash containing instruction of how to filter. Known keys:

=over 4

=item * C<re> (filter regex object)

With filter values: C<str> (stringify it).

=item * C<date> (date object, such as DateTime in Perl)

With filter values: C<epoch> (convert to Unix epoch time).

=item * C<code> (subroutine)

With filter values: C<str> (convert to string literal C<CODE>).

=back

More sophisticated filtering rules might be specified in the future.

Filtering using generated code can be faster since the code will use for() loop
and inline conversion instead of callback for each data item (higher subroutine
call overhead).

=for Pod::Coverage ^filter_using_.+

=head1 NOTES

Another alternative is using the encoder's facility, for example the L<JSON>
module can convert blessed object:

 use JSON 2;
 my $encoder = JSON->new->convert_blessed;
 {
     local *DateTime::TO_JSON = sub { $_[0]->ymd };
     print $encoder->encode($doc);
 }

But this currently can't convert coderefs. JSON also can't handle circular
references, which neither this wrapper property nor the above way can work
around.

=head1 SEE ALSO

L<Perinci>

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/Perinci-Sub-Property-result_postfilter>.

=head1 SOURCE

Source repository is at L<https://github.com/sharyanto/perl-Perinci-Sub-Property-result_postfilter>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=Perinci-Sub-Property-result_postfilter>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.013.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "(DEPRECATED) Postfilter function result",
  "AUTHOR" => "Steven Haryanto <stevenharyanto\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Perinci-Sub-Property-result_postfilter",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Perinci::Sub::Property::result_postfilter",
  "PREREQ_PM" => {
    "Perinci::Sub::PropertyUtil" => 0
  },
  "TEST_REQUIRES" => {
    "DateTime" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Perinci::Sub::Wrapper" => "0.57",
    "Test::More" => "0.98",
    "Test::Perinci::Sub::Wrapper" => "0.57"
  },
  "VERSION" => "0.08",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DateTime" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Perinci::Sub::PropertyUtil" => 0,
  "Perinci::Sub::Wrapper" => "0.57",
  "Test::More" => "0.98",
  "Test::Perinci::Sub::Wrapper" => "0.57"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



